qbft/spectest/tests/controller/latemsg/late_commit_past_instance.go

```diff
@@ -36,30 +39,96 @@ func LateCommitPastInstance() tests.SpecTest {
 				DecidedVal: testingutils.TestingQBFTFullData,
 				DecidedCnt: 1,
 			},
-			ControllerPostRoot: postRoot,
+			ControllerPostRoot:  sc.Root(),
+			ControllerPostState: sc.ExpectedState,
 		}
 	}
 
+	lateMsg := testingutils.TestingCommitMultiSignerMessageWithHeight([]*bls.SecretKey{ks.Shares[4]}, []types.OperatorID{4}, 4)
+	sc := lateCommitPastInstanceStateComparison(6, lateMsg)
+
 	return &tests.ControllerSpecTest{
 		Name: "late commit past instance",
 		RunInstanceData: []*tests.RunInstanceData{
-			instanceData(qbft.FirstHeight, "24cf697092529cfab3ab06b969d8696692c8bcbb9f41a954f71dc74c3b1d7e97"),
-			instanceData(1, "8aa5464b119518f178d81edf4cea1f4c918f9e084e5262a0e276d3afb00ba620"),
-			instanceData(2, "1799fe0981ae08bde1eae9fef88ef8035f5952974647786287a1a8c36544a5da"),
-			instanceData(3, "2ddf2b8c2f35c8115ddd68120c71e64809bfea6b023ed13e177c2474a95d137d"),
-			instanceData(4, "318816cc8819ad062996704fb4b9990b8088ade0cb4c26816ea0965783bff12e"),
-			instanceData(5, "fd83cdee705de628cf4a9baf9e662c424bf0942c63a680266d7db872d16e9f0a"),
+			instanceData(qbft.FirstHeight),
+			instanceData(1),
+			instanceData(2),
+			instanceData(3),
+			instanceData(4),
+			instanceData(5),
 			{
 				InputValue: []byte{1, 2, 3, 4},
 				InputMessages: []*qbft.SignedMessage{
-					testingutils.TestingCommitMultiSignerMessageWithHeight(
-						[]*bls.SecretKey{ks.Shares[4]},
-						[]types.OperatorID{4},
-						4,
-					),
+					lateMsg,
 				},
-				ControllerPostRoot: "691ac09194b0f6c67e85de702293f19863b0f5ee19e17c3764e411cee253f8b7",
+				ControllerPostRoot:  sc.Root(),
+				ControllerPostState: sc.ExpectedState,
 			},
 		},
 	}
 }
+
+func lateCommitPastInstanceStateComparison(height qbft.Height, lateMsg *qbft.SignedMessage) *qbftcomparable.StateComparison {
+	identifier := []byte{1, 2, 3, 4}
+	config := testingutils.TestingConfig(testingutils.Testing4SharesSet())
+	contr := testingutils.NewTestingQBFTController(
+		identifier[:],
+		testingutils.TestingShare(testingutils.Testing4SharesSet()),
+		config,
+	)
+
+	ks := testingutils.Testing4SharesSet()
+	allMsgs := testingutils.DecidingMsgsForHeightWithRoot(testingutils.TestingQBFTRootData,
+		testingutils.TestingQBFTFullData, testingutils.TestingIdentifier, 5, ks)
+
+	for i := 0; i <= int(height); i++ {
+		contr.Height = qbft.Height(i)
+		_ = contr.StartNewInstance([]byte{1, 2, 3, 4})
+
+		offset := 7 * i
+		msgs := allMsgs[offset : offset+7]
+
+		state := testingutils.BaseInstance().State
+		state.Height = qbft.Height(i)
+
+		// last height
+		if lateMsg != nil && i == int(height) {
+			contr.StoredInstances[0].State = state
+			break
+		}
```

Comment:
@olegshmuelov
Why do we have this condition? I see that it works but not sure why we skip the last height

Reply:
the last height has the late message only, so the state for this height is not decided

Reply:
We should check what's up with the instance at height 4 (since it is a latemsg)

Reply:
ok, it all looks good.
But maybe let's start making state default values explicit